tools/c7n_oci/c7n_oci/mu.py

```diff
@@ -0,0 +1,620 @@
+import json
+import logging
+import os
+import re
+from abc import ABC, abstractmethod
+
+import oci
+import yaml
+from c7n_oci.constants import COMPARTMENT_IDS
+from oci.events.models import ActionDetailsList, CreateFaaSActionDetails, CreateRuleDetails
+
+from c7n.utils import local_session
+from c7n.version import version
+
+log = logging.getLogger('custodian.oci.serverless')
+
+
+def is_resource_exists(resource):
+    return resource and resource[0].freeform_tags.get('SourceTool') == 'Custodian'
+
+
+class FunctionImageHandler:
+    def __init__(self, session, tenancy):
+        self.session = session
+        self.tenancy = tenancy
+
+    def initialize(self):
+        self.artificats_client = self.session.client('oci.artifacts.ArtifactsClient')
+        self.storage_client = self.session.client('oci.object_storage.ObjectStorageClient')
+        self.identity_client = self.session.client('oci.identity.IdentityClient')
+
+        # Importing here to avoid docker dependency at oracle function runtime
+        import docker
+
+        self.docker_client = docker.from_env()
+        self.repo_name = "cloudcustodian/c7n-oci"
+        self.repo_namespace = self._get_namespace()
+        self.oci_repo = f"{self._get_region_key()}.ocir.io"
+        self.ocir_repo_path = f"{self.oci_repo}/{self.repo_namespace}/{self.repo_name}"
+        self.tag = "latest"
+        self.ocir_image_name = f"{self.ocir_repo_path}:{self.tag}"
+        self.docker_hub_image_name = f"{self.repo_name}:{self.tag}"
+        self._oci_repo_login()
+
+    def _oci_repo_login(self):
+        response = self.identity_client.get_user(self.session.get_config()["user"])
+        self.docker_client.login(
+            username=f"{self.repo_namespace}/{response.data.name}",
+            password=os.environ['OCI_AUTH_TOKEN'],
+            registry=self.oci_repo,
+        )
+
+    def _get_namespace(self):
+        return self.storage_client.get_namespace().data
+
+    def _get_region_key(self):
+        region_key = None
+        data = self.identity_client.list_region_subscriptions(self.tenancy).data
+        for region in data:
+            if region.region_name == self.session.get_config()['region']:
+                region_key = region.region_key.lower()
+
+        if region_key:
+            return region_key
+        else:
+            raise ValueError(
+                f"Region {self.session.get_config()['region']} is not matching with \
+                    any subscribed region."
+            )
+
+    def pull(self):
+        log.debug("Pulling the image: %s", self.docker_hub_image_name)
+        self.docker_client.images.pull(self.docker_hub_image_name)
```

Comment:
we probably need a callback to log progress, docker image loads can be take a long time.

Reply:
Added code to check the progress.